File: 01_introduction_to_algorithms/ruby/binary_search.rb
# irb -r ./01_introduction_to_algorithms/ruby/binary_search.rb
class BinarySearch
  def self.search(list, number, low = 0, high = list.length - 1)
    middle = (low + high) / 2
    guess = list[middle]

    if guess == number
      return {position: middle, guess: guess}
    elsif guess > number
      high = middle - 1
      return self.search(list, number, low, high)
    else
      low = middle + 1
      return self.search(list, number, low, high)
    end
  rescue SystemStackError => error
    return nil
  end
end

File: 01_introduction_to_algorithms/elixir/binary_search.ex
defmodule BinarySearch do
  def search([], _number), do: nil
  def search(list, number) do
    search(list, number, 0, Enum.count(list))
  end

  defp search(list, number, low, high) do
    middle = div((low + high), 2)
    guess = Enum.at(list, middle)

    cond do
      guess == number ->
        [position: middle, guess: guess]

      guess > number ->
        high = middle - 1
        search(list, number, low, high)

      true ->
        low = middle + 1
        search(list, number, low, high)
    end
  end
end
